blib/libdoc/Psy.Psy.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.3
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Psy::Psy 3"
.TH Psy::Psy 3 "2005-11-15" "perl v5.8.7" "User Contributed Perl Documentation"
.SH "VERSION"
.IX Header "VERSION"
Version 0.01
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
A wrapper for various purposes.  Should contain base functions that are inherited by all modules (versions etc.).
.PP
Example usage: 
	my \f(CW%params\fR = ( 
		'\-matrix_file' => 'matrix',
		'\-matrix_label' => 'my_label',
		'\-helix_index' => 'stem_index',
		'\-path' => 'some/path/',
		'\-number_terminals' => 8,  
		'\-project_name' => 'test'
	);
.PP
.Vb 1
\&        my $foo = Psy->new(%params);
.Ve
.SH "EXPORT"
.IX Header "EXPORT"
.SH "OBJECT"
.IX Header "OBJECT"
.Sh "_verify"
.IX Subsection "_verify"
internal, does nothing thus far
.Sh "mx"
.IX Subsection "mx"
a matrix accessor
.SH "AUTHOR"
.IX Header "AUTHOR"
\&'Matt, \f(CW\*(C`<m{j}yoder@{tee}[aye](em)(you).domain4unis>\*(C'\fR
.SH "BUGS"
.IX Header "BUGS"
Please report any bugs or feature requests to
\&\f(CW\*(C`bug\-psy\-psy@rt.cpan.org\*(C'\fR, or through the web interface at
<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Psy>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.
.SH "ACKNOWLEDGEMENTS"
.IX Header "ACKNOWLEDGEMENTS"
.SH "COPYRIGHT & LICENSE"
.IX Header "COPYRIGHT & LICENSE"
Copyright 2005 'Matt, all rights reserved.
.PP
This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Psy

see svn


blib/lib/Psy/Psy.pm
package Psy;

use warnings;
use strict;
use Data::Dumper;
use Carp;


use Exporter;
use vars qw(@ISA $VERSION @EXPORT_OK $PSY_BASE_DIR);
@ISA = qw(Exporter);

# some library fu
BEGIN {
	use FindBin qw($Bin);		# sets $Bin to the root directory
	our $PSY_BASE_DIR = $Bin;	# this directory is the directory the script is executed from??
 	our $debug;

	#	eval { require Psy::Config; }; # won't fail if not there
	#	if ( ! $@ ) { # $@ 
			#		push @INC, "Templates";
			#	print @INC;
		}
	# if Psy::Config is not created (i.e. Psy not formally installed) then add the Templates directory to the path
	
	#}

use Psy::Matrix::Matrix;
use Psy::Dna::Iupac;
use Psy::Output::Output;
use File::Spec;
=head1 NAME

Psy::Psy

=head1 VERSION

Version 0.01

=cut

$VERSION = '0.01';

=head1 SYNOPSIS

A wrapper for various purposes.  Should contain base functions that are inherited by all modules (versions etc.).
 
Example usage: 
	my %params = ( 
		'-matrix_file' => 'matrix',
		'-matrix_label' => 'my_label',
		'-helix_index' => 'stem_index',
		'-path' => 'some/path/',
		'-number_terminals' => 8,  
		'-project_name' => 'test'
	);

	my $foo = Psy->new(%params);


=head1 EXPORT

=cut

# @EXPORT_OK = qw(new generatedByHeader version rootpath projectname starttime $ROOT_DIR $PSY_BASE_DIR $PSY_ROOT_DIR);

=head1 OBJECT

=cut

sub new {
	my $type = shift;
	my %params = @_;	
  	my $self = {};
	
	bless $self, $type;
	$self->_init(%params);

    return $self;                 
}

sub _init {
	my $self= shift;
	my %params = @_;
	
	print "\n";
	
	#srand();
	
	%params = $self->_verify(%params);

	# print "params after verify: ", %params, "\n";
	
	$self->{_matrix_accession} = 0;
	$self->{_structure_accession} = 0;
	
	$self->{matricies} = {}; # named hash pointing to matrices	
	$self->version($VERSION);
	$self->rootpath();
	$self->starttime();
	
	$self->projectname($params{'-project_name'});

	my $working_matrix;
	
	if (defined $params{'-matrix_file'}) {
	   	print "_init matricies from Psy.pm\n";	
		$working_matrix = $self->_init_matricies(%params)
	};

	if ($params{'-helix_index'}) {
		 $self->mx( '-matrix_label' => $working_matrix)->{structures}->{"original"}->_load(%params);  
	};

	if (($params{'-matrix_file'}) and ($params{'-helix_index'})) { 
			 $self->mx('-matrix_label' => $working_matrix)->{structures}->{"original"}->_build_helices(
																							'-mx' => $self->mx( '-matrix_label' => $working_matrix ),
																							%params
																						)
	}

	return 1;
}


=head2 _verify

internal, does nothing thus far

=cut

sub _verify {
	my $self=shift;
	my %params = @_;
		#	foreach my $p (keys %params) { # check Psy parameters here
		#			print "$p\n";
		#		if ($p eq '-datapath') {
		#	}
		#}

	$params{'-project_name'} = 'my_project' if not defined $params{'-projectname'};
	
	return %params;
}


=head2 mx

a matrix accessor

=cut

sub mx {
	my $self=shift;

	if (@_) {

		# debug
		# print "MX PARAMS: ";
		# print join " ", @_;
		# print "\n";
		
		my %params = @_;
		if ($params{'-matrix_label'}) {		 
			return ( $self->{matricies}->{$params{'-matrix_label'}} )
		}

		# run a test on the mx object here
	}
	else {
		croak "no matrix requested"
	}	
	0
#	return {shift $self->{matricies} } # no matrix requested, return a reference to the last in stack
}

sub version {
	my $self = shift;
	if (@_) { $self->{VERSION} = shift; }
	return $self->{VERSION};
}

sub rootpath {
	my $self = shift;
	if (not defined $self->{ROOTPATH}) { use FindBin qw($Bin); $self->{ROOTPATH} = $Bin; }
	return $self->{ROOTPATH};
}

sub starttime {
	my $self = shift;
	if (not defined $self->{STARTTIME}) { $self->{STARTTIME} = localtime($^T); }
	return $self->{STARTTIME};
}

sub projectname {
	my $self = shift;
	if (@_) { $self->{PROJECTNAME} = shift; }
	return $self->{PROJECTNAME};
}

sub generatedByHeader {
	my $self = shift;
	my $h = "generated by Psy ver. ", $self->version, " on ", $self->starttime, " for project ", $self->projectname;
}

sub _init_matricies {
	my $self=shift;
	my %params = @_;
	my $label;

 	print "incoming label:",	$params{'-matrix_label'} , "\n";; 
	
	if (defined	$params{'-matrix_label'}) {
		$label = $params{'-matrix_label'}
	}	
	else {
		$label = "matrix_".$self->{_matrix_accession}; 
	}

	$self->{matricies}->{$label} = matrix->new(%params);
	
	$self->{_matrix_accession}++;

	# handles pointers to hashes as such -> single data path-> appended to all; hash -> one to many, no path, array - many

	print "\ninitialized matrix: $label\n";

	return ($label); # necessary!
}

sub _init_helices {
	my $self=shift;
	my %params = @_;

	# required:
	# -mx => matrix object
	
	# optional:
	# -helix_slice => a slice object
	
	croak "no matrix" if not $params{'-matrix_label'};
	my $mx =  $self->mx(%params);
	
	my $slice;

	if (defined $params{'-helix_slice'}) { 
	   	$slice = $params{'-helix_slice'} 
	} 
   	else {
		$slice = $mx->orig_slice()
	};

   	print "slice:\n", $slice->decribe_coded();
	print "\n";
	
	$self->mx(%params)->{structures}->{"original"}->_build_helices(
																	'-mx' => $self->mx(%params),
																	'-helix_slice' => $slice
																);
}

=head1 AUTHOR

'Matt, C<< <m{j}yoder@{tee}[aye](em)(you).domain4unis> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-psy-psy@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Psy>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2005 'Matt, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Psy::Psy
